src/analyses/dependence_graph.cpp

```diff
@@ -67,42 +53,54 @@ void dep_graph_domaint::control_dependencies(
 {
   // Better Slicing of Programs with Jumps and Switches
   // Kumar and Horwitz, FASE'02:
-  // Node N is control dependent on node M iff N postdominates one
-  // but not all of M's CFG successors.
+  // "Node N is control dependent on node M iff N postdominates, in
+  // the CFG, one but not all of M's CFG successors."
   //
-  // candidates for M are from and all existing control-depended on
-  // nodes; from is added if it is a goto or assume instruction
-  if(from->is_goto() ||
-     from->is_assume())
-    control_deps.insert(from);
+  // The "successor" above refers to an immediate successor of M.
+
+  // Candidates for M for "to" are "from" and all existing control
+  // dependencies on nodes. "from" is added if it is a goto or assume
+  // instruction
+
+  // Add new candidates
+
+  if(from->is_goto() || from->is_assume())
+    control_dep_candidates.insert(from);
+  else if(from->is_end_function())
+  {
+    control_dep_candidates.clear();
+    return;
+  }
+
+  if(control_dep_candidates.empty())
+    return;
+
+  // Get postdominators
 
   const irep_idt id=from->function;
   const cfg_post_dominatorst &pd=dep_graph.cfg_post_dominators().at(id);
 
-  // check all candidates for M
-  for(depst::iterator
-      it=control_deps.begin();
-      it!=control_deps.end();
-      ) // no ++it
-  {
-    depst::iterator next=it;
-    ++next;
+  // Check all candidates
 
-    // check all CFG successors
+  for(const auto &cd : control_dep_candidates)
```

Comment:
I think this bit is just cleanup for clarity and not a substantial change? If so, if not already, move it into a different commit than the substantial change.

Reply:
The set `control_dep_candidates` was introduced in the fix so this is new code.